Fix URL patterns for videos and thumbnails

andersonvom opened the pull request:
Some image URLs on latelelibre start with either '//latelelibre.fr/...' or just '/wp-content/...'. This makes sure that all thumbnail URLs are sanitized before being passed along to Kodi so that they all load correctly.

It also fixes the playable video URL pattern, since it used to assume that attributes within the quality array would always come back in the same order. Since this is not the case, it now searches within each array for the "url" attribute alone, ignoring all other attributes.

andersonvom committed:
Fix URL patterns for series thumbnails

Some URLs on latelelibre start with either '//latelelibre.fr/...' or
just '/wp-content/...'. This makes sure that all thumbnail URLs are
sanitized before being passed along to Kodi so that they all load
correctly.

andersonvom committed:
Fix whitespace in all python files

andersonvom committed:
Sort video resolution in descending order of quality

This adds 1080p and 144p as supported resolutions for dailymotion and
sorts them in descending order of quality for better eperience.

andersonvom committed:
Fix playable video URL pattern

The previous pattern assumes that the JSON blob in the embed video page
always comes back with attributes in the same order, but this is not the
case

Instead it now searches within each video quality array for anything
that's not the end of the array followed by a "url" attribute. It also
uses the `*?` [1] patterns for non-greedy matches, thus simplifying the
regexp as a whole.

[1]: https://docs.python.org/2/howto/regex.html#greedy-versus-non-greedy

jamontes commented on resources/lib/ltl_api.py:
```diff
@@ -417,7 +417,7 @@ def get_playable_youtube_url(video_id):
 def get_playable_dailymotion_url(video_id):
     """This function returns the playable URL for the Dalymotion embedded video from the video_id retrieved."""
-    daily_video_pattern = '"%s":\[{"type":"video\\\/mp4","url":"(.+?)"'
+    daily_video_pattern = '"%s":\[[^]]*,"url":"(.+?)"'
```
Your modified regexp it's taking now the last URL option from the JSON array, which incidentally is the good one (i.e. the mp4 file link), but it would fail in the future if the options order into the JSON array varies (as it did this last time with my broken regexp). In order to make the regexp array order independent, please consider this one instead into your code modification:

`video_quality_pattern = '"%s":\[[^]]*{"type":"video\\\/mp4","url":"([^"]*)"'`

This regexp would guarantee that the option chosen is the good one, and skip over any other undesired options, no matters of the order or number of options available into the array for each quality option.

Thanks a lot for your help and effort,

jamontes.

andersonvom replied:
Oh I see what you mean. Good catch!
I'll make these changes and submit them shortly.

jamontes commented:
Hey andersonvom,

Thanks a lot for contributing to this add-on and take care of the code like yours!!

I appreciate very much all the effort and time you have spent in cleaning up and fixing everything to make it working again, and I would be glad to merge your pull request with the four commits into the master branch, but your modified regexp pattern should be reviewed before merge (please, see: [r104145161](https://github.com/jamontes/plugin.video.latelelibre_fr/pull/2/commits/de3c520d73d2c75cd990d11cb6aca3b3391a9ab3#r104145161))

Please, consider to review and modify the regexp as proposed, and I'll be happy to merge your PR into the master branch.

Thanks a lot for all your help and effort.

Best regards,

jamontes.

jamontes reviewed (changes requested):
Please, have a look at the comments about the regexp pattern. Thanks!

jamontes reviewed (changes requested):
Hey andersonvom,

That's exactly the regexp I made, when started fixing the add-on on my own, just before receive your first commit a few days ago!

Then I've received your PR and reviewed the changes you have made into the code in order to validate them, testing your regexp with some pattern examples.

Then I realized that your regexp wasn't valid, nor the mine either! Applying the current regexp as is, over a quality without mp4 link, would jump up to the next quality item and pick up the link of the next, associating it to the previous missing one (i.e. if quality "380" has missing the mp4 link, the regexp would pick up the mp4 link related to the "480" quality, if it's present).

By using the `'[^]]*?'` instead of `'.*?'` into the regexp, you'll make it sure it's going to pick up the right link associated with its quality. By testing further the regexp with more examples, I contemplated as well the case of the empty URL, so it led to this final version of the regexp:

`daily_video_pattern = '"%s":\[[^]]*?{"type":"video\\\/mp4","url":"([^"]+?)"'`

Thus it takes the non-greedy match into account, as well as the mp4 missing link and the empty URL link exceptions, making the regexp more robust than ever.

Please, consider this final (I hope) version of the regexp into your commit.

Many thanks for your help and best regards,

jamontes.

jamontes commented on resources/tests/test_api.py:
```diff
@@ -30,21 +30,21 @@ def test_1_create_index(self):
     def test_2_video_docs(self):
         doc_list = api.get_video_docs()
-        self.assertTrue(doc_list is not None and len(doc_list) > 10)
+        self.assertEqual(len(doc_list), 12)
```
Please, leave it as is. This condition is website content dependent and will vary with the time.

andersonvom replied:
The test as it is is broken (you can try running it yourself). We should probably come up with a better assertion.

jamontes commented on resources/tests/test_api.py:
```diff
@@ -30,21 +30,21 @@ def test_1_create_index(self):
         menu_url = 'http://latelelibre.fr/emissions/thibault-o-tour-du-monde/'
         video_list = api.get_video_sec(menu_url)
-        self.assertTrue(video_list is not None and len(video_list) > 8)
+        self.assertEqual(len(video_list), 12)
```
Please, leave it as is. This condition is website content dependent and will vary with the time.

jamontes commented on resources/tests/test_api.py:
```diff
@@ -30,21 +30,21 @@ def test_1_create_index(self):
         menu_url = 'http://latelelibre.fr/series/une-bien-belge-histoire/'
         video_list = api.get_video_sec(menu_url)
-        self.assertTrue(video_list is not None and len(video_list) > 4)
+        self.assertEqual(len(video_list), 5)
```
Please, leave it as is. This condition is website content dependent and will vary with the time.

jamontes commented on resources/tests/test_api.py:
```diff
@@ -30,21 +30,21 @@ def test_1_create_index(self):
     def test_3_video_items(self):
         video_items = api.get_video_items()
-        self.assertTrue(len(video_items.get('video_list', '')) > 15)
+        self.assertEqual(len(video_items.get('video_list', '')), 15)
```
Here, the same criteria should be applied as with the others; the video list is website dependent and its quantity is expected to vary along the time (as it actually did between your changes). As it's sole purpose is to test that a significant valid list of videos is retrieved, testing a quantity above 10 (i.e. 10 videos + the "next page" entry) should be enough. Using an exact match assert type will led to failed error tests in future commits.

So please, consider to change the line like this:

`self.assertTrue(len(video_items.get('video_list', '')) > 10)`

This line should be the only one it should be modified from my point of view in this commit.

Thanks!

andersonvom replied:
OK! =)

jamontes reviewed (changes requested):
Hi andersonvom,

Thanks a lot for point me out your approach about the tests!

I agree with you that they are something brittle and conditioned to changes from the website. That's the reason why they avoid the exact match (i.e. assertEqual) and tries the minimum list results approach, with a conservative quantity that guaranties somehow the API isn't broken completely (or at least it retrieves something).

So the values used on every test are expected to be conservative ones, and my failure on test 3 was to think the video list would never lower the quantity seen as far as now (this value isn't content dependent, but website one).

Please, consider to leave the rest of the test unchanged, and setup the value of test 3 to a lower one (10 would be perfect) for this commit.

Thanks a lot for your help and effort,

jamontes.

andersonvom committed:
Fix tests that broke due to change in website

This test was already broken before the recent changes due to changes in
the API response.  As it is, it's still brittle since every time the
response changes, it could potentially break.

Going forward, if it keeps breaking, we should probably mock the API
response and maybe add contract tests to verify that the website
responds as expected.

andersonvom commented:
Yeah, I see your point and it's, indeed, tricky.
I've reverted the tests back, changing just the one that was broken.
Hopefully it should be good to go now.! =)

--
Anderson Mesquita

On Sat, Mar 4, 2017 at 2:44 PM, jamontes <notifications@github.com> wrote:

> *@jamontes* requested changes on this pull request.
>
> Hi andersonvom,
>
> Thanks a lot for point me out your approach about the tests!
>
> I agree with you that they are something brittle and conditioned to
> changes from the website. That's the reason why they avoid the exact match
> (i.e. assertEqual) and tries the minimum list results approach, with a
> conservative quantity that guaranties somehow the API isn't broken
> completely (or at least it retrieves something).
>
> So the values used on every test are expected to be conservative ones, and
> my failure on test 3 was to think the video list would never lower the
> quantity seen as far as now (this value isn't content dependent, but
> website one).
>
> Please, consider to leave the rest of the test unchanged, and setup the
> value of test 3 to a lower one (10 would be perfect) for this commit.
>
> Thanks a lot for your help and effort,
>
> jamontes.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jamontes/plugin.video.latelelibre_fr/pull/2#pullrequestreview-25130873>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAERIrKHU2Zshby21zqbTLerj2pxp1Xdks5rib8hgaJpZM4MOlev>
> .
>

jamontes commented:
Yes, it's ready to go now!

Thanks a lot for your help and contribution.

Best regards,

jamontes.

andersonvom commented:
Anytime! =)
